dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java

```diff
@@ -0,0 +1,424 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.dxf2.events.importer.context;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import lombok.Data;
+
+import org.apache.commons.collections.MapUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.cache.Cache;
+import org.hisp.dhis.cache.CacheProvider;
+import org.hisp.dhis.category.CategoryCombo;
+import org.hisp.dhis.common.IdScheme;
+import org.hisp.dhis.common.IdSchemes;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.dxf2.common.ImportOptions;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.organisationunit.FeatureType;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramStageDataElement;
+import org.hisp.dhis.program.ProgramType;
+import org.hisp.dhis.program.ValidationStrategy;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.user.sharing.Sharing;
+import org.springframework.core.env.Environment;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+/**
+ * This supplier builds and caches a Map of all the Programs in the system.
+ * For each Program, the following additional data is retrieved:
+ *
+ *
+ *
+ * @formatter:off
+ *
+         Program
+         +
+         |
+         +---+ Program Stage (List)
+         |           |
+         |           +---+ User Access (ACL)
+         |           |
+         |           +---+ User Group Access (ACL)
+         |
+         |
+         +---+ Category Combo
+         |
+         |
+         +---+ Tracked Entity Instance
+         |           |
+         |           +---+ User Access (ACL)
+         |           |
+         |           +---+ User Group Access (ACL)
+         |
+         |
+         |
+         +---+ Organizational Unit (List)
+         |
+         +---+ User Access (ACL)
+         |
+         +---+ User Group Access (ACL)
+ * @formatter:on
+ *
+ * @author Luciano Fiandesio
```

Comment:
Identity theft is not a joke, Jim! Millions of families suffer every year!

Reply:
I didn't create this file, just minor updates for this issue